tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc

```diff
@@ -254,68 +227,70 @@ bool CouldNeedLibdevice(const llvm::Module& module) {
-  }
+// Links the module with a vector of path to bitcode modules
+// The paths are guaranteed to exist.
+Status LinkWithBitcodeVector(llvm::Module* module,
+                             const std::vector<string>& bitcode_path_vector) {
+  llvm::Linker linker(*module);
 
-  VLOG(1) << "Linking with libdevice from: " << libdevice_path;
-  std::unique_ptr<llvm::Module> libdevice_module =
-      LoadIRModule(libdevice_path, &module->getContext());
+  for (auto& bitcode_path : bitcode_path_vector) {
+    if (!tensorflow::Env::Default()->FileExists(bitcode_path).ok()) {
+      LOG(WARNING) << "bitcode module is required by this HLO module but was "
+                      "not found at "
+                   << bitcode_path;
+      return xla::InternalError("bitcode module not found at %s", bitcode_path);
+    }
 
-  llvm::Linker linker(*module);
-  if (linker.linkInModule(
-          std::move(libdevice_module), llvm::Linker::Flags::LinkOnlyNeeded,
-          [](Module& M, const StringSet<>& GVS) {
-            internalizeModule(M, [&GVS](const GlobalValue& GV) {
-              return !GV.hasName() || (GVS.count(GV.getName()) == 0);
-            });
-          })) {
-    return xla::InternalError("Error linking libdevice from %s",
-                              libdevice_path);
+    std::unique_ptr<llvm::Module> bitcode_module =
+        LoadIRModule(bitcode_path, &module->getContext());
+    if (linker.linkInModule(
+            std::move(bitcode_module), llvm::Linker::Flags::LinkOnlyNeeded,
+            [](Module& M, const StringSet<>& GVS) {
+              internalizeModule(M, [&M, &GVS](const GlobalValue& GV) {
+                return !GV.hasName() || (GVS.count(GV.getName()) == 0);
+              });
+            })) {
+      return xla::InternalError("Error linking bitcode module from %s",
+                                bitcode_path);
+    }
   }
   return Status::OK();
 }
 
-StatusOr<string> CompileModuleToPtx(llvm::Module* module,
-                                    std::pair<int, int> compute_capability,
+StatusOr<std::unique_ptr<llvm::TargetMachine>>
+ConstructLLVMTargetMachineForModule(llvm::Module* module,
+                                    GpuVersion gpu_version,
                                     const HloModuleConfig& hlo_module_config,
-                                    const string& libdevice_dir_path) {
-  // If the module has no functions or globals, there's nothing to compile. Just
-  // return an empty string.
-  if (module->empty() && module->global_empty()) {
-    VLOG(2) << "Module '" << module->getName().str()
-            << "' is empty. Skipping compilation.";
-    return string();
+                                    const string& device_bitcode_dir_path) {
+  // Check if we are running the backend for NVPTX or AMDGPU
+  llvm::Triple target_triple = llvm::Triple(module->getTargetTriple());
+  bool isNVPTX = target_triple.isNVPTX();
+
+  if (isNVPTX) {
```

Comment:
if (isNVPTX) {
  ...
} else {
  ...
}
Is a highly repetitive pattern in this function. Please consider to fork the it into a NVPTX and AMD specific versions to improve readability and reduce the level of nesting.

Reply:
@thomasjoerg I think this is a design guideline I'd like to seek your opinion, as this request would also impact the frontend (`nvptx_compiler.cc`).

In the downstream fork I've been working on, originally we kept NVPTX and AMDGPU as separate modules, and they are enabled/disabled at configure-time via bazel `--config=` option. There were quite a few duplicated codes and @jlebar recommended unifying the logic as much as possible, hence I created this PR as a first step to make this `nvptx_backend_lib` to support both NVPTX and AMDGPU. And therefore I introduced this `isNVPTX` runtime check inside the function.

Based on your comments, would you think it'd be a good idea if I:

a) rename this function from `ConstructLLVMTargetMachine` to something like `LinkAndOptimizeModule`.
b) have two versions of `LinkAndOptimizeModule`, one in nvptx namespace, one in amdgpu namespace. And there would be some code duplication between the two versions, while keeping platform-specific logic separately.

Reply:
@thomasjoerg I've submitted a new version to address my points above. Please kindly have another round of review and see if you are okay with it. Thanks!